include/exadg/incompressible_navier_stokes/postprocessor/postprocessor.cpp

```diff
@@ -195,116 +237,142 @@ PostProcessor<dim, Number>::do_postprocessing(VectorType const &     velocity,
-  }
-
-
-  // streamfunction
-  if(pp_data.output_data.write_streamfunction)
-  {
-    streamfunction.type        = SolutionFieldType::scalar;
-    streamfunction.name        = "streamfunction";
-    streamfunction.dof_handler = &navier_stokes_operator->get_dof_handler_u_scalar();
-    navier_stokes_operator->initialize_vector_velocity_scalar(
-      streamfunction.get_vector_reference());
-    streamfunction.recompute_solution_field = [&](VectorType & dst, VectorType const &) {
-      navier_stokes_operator->compute_streamfunction(dst, evaluate_get(vorticity));
-    };
-
-    this->additional_fields_vtu.push_back(&streamfunction);
+    velocity_magnitude.reinit();
   }
 
   // q criterion
   if(pp_data.output_data.write_q_criterion)
   {
-    q_criterion.type        = SolutionFieldType::scalar;
-    q_criterion.name        = "q_criterion";
-    q_criterion.dof_handler = &navier_stokes_operator->get_dof_handler_u_scalar();
-    navier_stokes_operator->initialize_vector_velocity_scalar(q_criterion.get_vector_reference());
+    q_criterion.type              = SolutionFieldType::scalar;
+    q_criterion.name              = "q_criterion";
+    q_criterion.dof_handler       = &navier_stokes_operator->get_dof_handler_u_scalar();
+    q_criterion.initialize_vector = [&](std::shared_ptr<VectorType> & dst) {
+      dst = std::make_shared<VectorType>();
+      navier_stokes_operator->initialize_vector_velocity_scalar(*dst);
+    };
     q_criterion.recompute_solution_field = [&](VectorType & dst, VectorType const & src) {
       navier_stokes_operator->compute_q_criterion(dst, src);
     };
 
-    this->additional_fields_vtu.push_back(&q_criterion);
+    q_criterion.reinit();
   }
 
   // mean velocity
   if(pp_data.output_data.mean_velocity.is_active)
   {
-    mean_velocity.type        = SolutionFieldType::vector;
-    mean_velocity.name        = "mean_velocity";
-    mean_velocity.dof_handler = &navier_stokes_operator->get_dof_handler_u();
-    navier_stokes_operator->initialize_vector_velocity(mean_velocity.get_vector_reference());
-    mean_velocity.recompute_solution_field = [&](VectorType & dst, VectorType const & src) {
-      compute_mean_velocity(dst, src);
+    mean_velocity_vector = std::make_shared<VectorType>();
+    navier_stokes_operator->initialize_vector_velocity(*mean_velocity_vector);
+
+    mean_velocity.type              = SolutionFieldType::vector;
+    mean_velocity.name              = "mean_velocity";
+    mean_velocity.dof_handler       = &navier_stokes_operator->get_dof_handler_u();
+    mean_velocity.initialize_vector = [&](std::shared_ptr<VectorType> & dst) {
+      dst = mean_velocity_vector;
     };
 
-    this->additional_fields_vtu.push_back(&mean_velocity);
+    mean_velocity.reinit();
   }
 
   // cfl
   if(pp_data.output_data.write_cfl)
   {
-    cfl_vector.type                     = SolutionFieldType::cellwise;
-    cfl_vector.name                     = "cfl_relative";
```

Comment:
I wonder why there has not been an initialization of the vector in the previous/current code version?

Reply:
I was wondering about the same: https://github.com/exadg/exadg/pull/245#discussion_r996444168